cksum: accept non-UTF-8 filenames

BenWiederhake opened the pull request:
Fixes #6574.

Example:

```console
$ touch $'funky\xffname'
$ cargo run -q cksum $'funky\xffname'
4294967295 0 funky�name
$ cksum $'funky\xffname'
4294967295 0 funky�name
```

BenWiederhake commented:
CI failures are unrelated issues:
- #6534
- #6570 (both)
- And some new MacOS issue? Similar to #6145, but not identical? See below.

```
---- test_stat::test_stdin_pipe_fifo1 stdout ----
run: /Users/runner/work/coreutils/coreutils/target/aarch64-apple-darwin/debug/coreutils stat -
thread 'test_stat::test_stdin_pipe_fifo1' panicked at 'Expected stderr to be empty, but it's:
stat: cannot stat '-': Bad file descriptor (os error 9)
', tests/by-util/test_stat.rs:279:10
```

BenWiederhake commented:
Changes since initial push:
- Added a test (doink!)
- Changed the output method so that non-UTF-8 filenames are correctly output (without Unicode replacement encoding or other loss)
- Unified the "OsStr(ing) to bytes" conversion with `cut`, as already suggested in #6563

BenWiederhake commented:
Changes since last push description:
- Formatting (looks like something disabled my pre-commit hook, whoops)
- Added missing `#[cfg(unix)]` on `use`
- Fix the unified definition to work on windows – turns out, a `if cfg!()` branch must successfully compile even if it is not taken.
- Added some spellcheck exceptions. (These should really be added to the pre-commit hook)
- Restrict the test to linux-only, because MacOS doesn't allow non-UTF-8 filenames in the first place

BenWiederhake commented:
Changes since last push description:
- Now that the test is restricted, the initial import must only happen on linux. Aaargh!
- Both of them.

Remaining CI failures are unrelated:
- #6534
- #6570 (both)
- Timeout while downloading a 877 byte(!) file, probably because there have been so many restarts

BenWiederhake commented:
Changes since last push description:
- Rebase (also solves most CI failures, except #6534 of course)
- Use the shared code from #6563 for converting a `OsStr` to `&[u8]`
- Remove the `.to_os_string()` that's now unnecessary (This took two attempts – why didn't my clippy catch that?!)

cakebaker commented on tests/by-util/test_cksum.rs:
```diff
@@ -1250,3 +1250,30 @@ fn test_several_files_error_mgmt() {
+    let at = &scene.fixtures;
+    let filename: OsString = OsStringExt::from_vec(b"funky\xffname".to_vec());
+
+    at.touch(filename.clone());
```
A detail: instead of `clone()` you can simply use a reference. This also applies to lines 1268 and 1275.
```suggestion
    at.touch(&filename);
```

cakebaker commented on src/uu/cksum/src/cksum.rs:
```diff
@@ -116,52 +117,78 @@ where
-                "{} {}",
-                sum.parse::<u16>().unwrap(),
-                div_ceil(sz, options.output_bits)
+        let (before_filename, print_filename, after_filename) = match (options.algo_name, not_file)
```
The `print_filename` var name is a bit misleading and I was surprised it's of type `bool`. I expected it to be a `String`, like the other two vars. My suggestion is to rename it to something like `is_printable_filename`.

cakebaker commented on src/uu/cksum/src/cksum.rs:
```diff
@@ -116,52 +117,78 @@ where
+                    "{} {}",
+                    sum.parse::<u16>().unwrap(),
+                    div_ceil(sz, options.output_bits)
+                ),
+                false,
+                String::new(),
             ),
-            (ALGORITHM_OPTIONS_SYSV, false) => println!(
-                "{} {} {}",
-                sum.parse::<u16>().unwrap(),
-                div_ceil(sz, options.output_bits),
-                filename.display()
+            (ALGORITHM_OPTIONS_SYSV, false) => (
+                format!(
+                    "{} {} ",
+                    sum.parse::<u16>().unwrap(),
+                    div_ceil(sz, options.output_bits)
+                ),
+                true,
+                String::new(),
```
The difference between the two arms is difficult to see (a space at the end of the format string), though currently I don't see how this could be improved.

RenjiSann replied:
Maybe like this ?
```rust
(ALGORITHM_OPTIONS_SYSV, not_file) => {
    let mut s = format!(
        "{} {} ",
        sum.parse::<u16>().unwrap(),
        div_ceil(sz, options.output_bits)
    );

    // If file, remove the last space.
    if !not_file {
        // Assume all chars are ASCII.
        s.truncate(s.len() - 1);
    }

    (s, false, String::new())
}
```

BenWiederhake replied:
I solved it by removing `notfile` from the `match` structure entirely, and instead pushing it into the `format` for `before_filename`.

BenWiederhake commented:
Changes since (my?) last push:
- Remove `notfile` from the match structure, reducing code duplication
- Rename `print_filename` to `should_print_filename`
- Reduce filename copying in the test

BenWiederhake commented:
Changes since last push:
- Turn `if !not_file { " " } else { "" }` into `if not_file { "" } else { " " }` to make clippy happy (why wasn't this caught by pre-commit?)

cakebaker commented:
Thanks :)
